chain/network/src/concurrency/scope/tests.rs

```diff
@@ -0,0 +1,306 @@
+use crate::concurrency::ctx::Ctx;
+use crate::concurrency::scope;
+use crate::testonly::abort_on_panic;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+
+// drop a trivial future without completion => panic (in debug mode at least).
+#[tokio::test]
+#[should_panic]
+async fn must_complete_should_panic() {
+    let _ = scope::must_complete(async move { 6 });
+}
+
+// run a trivial future until completion => OK
+#[tokio::test]
+async fn must_complete_ok() {
+    assert_eq!(5, scope::must_complete(async move { 5 }).await);
+}
+
+#[tokio::test]
+async fn test_drop_service() {
+    abort_on_panic();
+    let res = scope::run!(&Ctx::inf(), |s| move |ctx: Ctx| async move {
+        let service = Arc::new(s.new_service());
+        service
+            .spawn(|ctx: Ctx| async move {
+                ctx.canceled().await;
+                Err(2)
+            })
+            .unwrap();
+        // Both scope task and service task await context cancellation.
+        // However, scope task drops the only reference to service before
+        // awaiting, which should cause service to cancel.
+        // The canceled service task returns an error which should cancel
+        // the whole scope.
+        drop(service);
+        ctx.canceled().await;
+        Ok(())
+    });
+    assert_eq!(Err(2), res);
+}
+
+#[tokio::test]
+async fn test_spawn_after_cancelling_scope() {
+    abort_on_panic();
+    let res = scope::run!(&Ctx::inf(), |s| move |ctx: Ctx| async move {
+        s.spawn(|_| async { Err(7) });
+        ctx.canceled().await;
+        s.spawn(|_| async { Err(3) });
+        Ok(())
+    });
+    assert_eq!(Err(7), res);
+}
+
+#[tokio::test]
+async fn test_spawn_after_dropping_service() {
+    abort_on_panic();
+    let res = scope::run!(&Ctx::inf(), |s| move |ctx: Ctx| async move {
+        let service = Arc::new(s.new_service());
+        service
+            .spawn({
+                let service = service.clone();
```

Comment:
I've delegated all refcounting to Arc to not trigger cleanup logic manually.
Sure, I can wrap Arc<Service> into yet another type. FYI these tests look a bit ugly because they are testing a specific lifecycle of the service object, but in the actual application it will be way more straightforward.